Userland/Libraries/LibCore/Stream.h

```diff
@@ -0,0 +1,838 @@
+    //        AK::CircularDuplexBuffer inlines its entire contents, and that
+    //        would make for a very large object on the stack.
+    //
+    //        The proper fix is to make a CircularQueue which uses a buffer on
+    //        the heap.
+    ByteBuffer m_buffer;
+    size_t m_buffered_size { 0 };
+};
+
+// NOTE: A Buffered which accepts any Stream could be added here, but it is not
+//       needed at the moment.
+
+template<IsSeekableStream T>
+class BufferedSeekable final : public SeekableStream {
+    friend BufferedHelper<T>;
+
+public:
+    static KResultOr<BufferedSeekable<T>> create(T&& stream, size_t buffer_size = 16384)
+    {
+        return BufferedHelper<T>::template create_buffered<BufferedSeekable>(move(stream), buffer_size);
+    }
+
+    BufferedSeekable(BufferedSeekable&& other) = default;
+    BufferedSeekable& operator=(BufferedSeekable&& other) = default;
+
+    virtual bool is_readable() const { return m_helper.stream().is_readable(); }
+    virtual KResultOr<size_t> read(Bytes buffer) override { return m_helper.read(move(buffer)); }
+    virtual bool is_writable() const { return m_helper.stream().is_writable(); }
+    virtual KResultOr<size_t> write(ReadonlyBytes buffer) { return m_helper.stream().write(buffer); }
+    virtual bool is_eof() const override { return m_helper.is_eof(); }
+    virtual bool is_open() const { return m_helper.stream().is_open(); }
+    virtual void close() { m_helper.stream().close(); }
+    virtual KResultOr<off_t> seek(i64 offset, SeekMode mode) override
+    {
+        auto result = TRY(m_helper.stream().seek(offset, mode));
+        m_helper.clear_buffer();
+        return result;
+    }
+
+    KResultOr<size_t> read_line(Bytes buffer) { return m_helper.read_line(move(buffer)); }
+    KResultOr<size_t> read_until(Bytes buffer, StringView const& candidate) { return m_helper.read_until(move(buffer), move(candidate)); }
+    template<size_t N>
+    KResultOr<size_t> read_until_any_of(Bytes buffer, Array<StringView, N> candidates) { return m_helper.read_until_any_of(move(buffer), move(candidates)); }
+    KResultOr<bool> can_read_line() { return m_helper.can_read_line(); }
+
+    size_t buffer_size() const { return m_helper.buffer_size(); }
+
+    virtual ~BufferedSeekable() override { }
+
+private:
+    BufferedSeekable(T stream, ByteBuffer buffer)
+        : m_helper(Badge<BufferedSeekable<T>> {}, move(stream), buffer)
+    {
+    }
+
+    BufferedHelper<T> m_helper;
+};
+
+template<IsSocket T>
+class BufferedSocket final : public Socket {
+    friend BufferedHelper<T>;
+
+public:
+    static KResultOr<BufferedSocket<T>> create(T&& stream, size_t buffer_size = 16384)
+    {
+        return BufferedHelper<T>::template create_buffered<BufferedSocket>(move(stream), buffer_size);
+    }
+
+    BufferedSocket(BufferedSocket&& other)
+        : Socket(static_cast<Socket&&>(other))
+        , m_helper(move(other.m_helper))
+    {
+        setup_notifier();
```

Comment:
Note about notifiers and buffering:
- the buffering object slurps up all the data, hitting EOF
- underlying socket's notifier does _not_ fire anymore, since there's no further data
- we now have data that is potentially never read, possible deadlock or spin.

Reply:
Yeah, the situation with notifiers and EOF is very error-prone. I can't quite think of a solution that doesn't spin, because EOF sockets are always available to read until closed. 

There needs to be a way to determine whether a socket is EOF by Notifier. That _could_ be through giving it a `did_eof` callback sort of thing, where the notifier disables itself after firing one last read notification through it. However, that also feels kind of half-baked.

Note that closing a Stream API socket will immediately disable the notifier before doing anything else, which avoids a lot of the spinning that was observed.

Reply:
Well, I think LibHTTP is gonna be a little...difficult to port now (it does have some assumptions about the behaviours of `Core::*`, but I think I've documented them all so they should be simple to spot)

Anyway, this is just a note, I hope we manage to resolve this in the future, because the workarounds are quite...sad.

Reply:
Here's the solution that has been reached during discussion on Discord:

1. `Notifier` gains a method that makes it send one last `on_ready_to_read` notification before disabling itself.
2. When EOF has been reached, socket subclasses call this method.
3. The closing of the socket is still manually done by its users; however, the Notifier spam will no longer happen even if the socket isn't closed, because the Notifier is no longer enabled.

Some example code for a user of the socket API with a notifier now becomes:

```cpp
socket.on_ready_to_read = [&] {
    auto maybe_nread = socket.read(some_buffer);
    // handle error here

    if (socket.is_eof()) {
        socket.close();
        return;
    }

    // the socket wasn't EOF and we received data;
    // process the received data as usual
}
```